### PR TITLE
adding reverse openai gym wrapper and tests

### PR DESCRIPTION
**Pull request by alexminnaar**
This PR is related to https://github.com/deepmind/bsuite/issues/6 where I suggested a wrapper that converts Openai gym environments to bsuite environments. This PR contains the wrapper and a test case that converts the Openai CartPole-v0 environment to a bsuite environment.

### CODE REVIEW

File: bsuite/utils/gym_wrapper.py [hunk 1]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
   def __getattr__(self, attr):
     """Delegate attribute access to underlying environment."""
     return getattr(self._env, attr)
+
+
+class ReverseGymWrapper(dm_env.Environment):
+    """A wrapper that converts an OpenAI gym environment to a dm_env.Environment"""
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Looks like you're using 4 spaces here. Are you able to switch to 2 to be consistent with the rest of the codebase?

File: bsuite/utils/gym_wrapper.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 # you may not use this file except in compliance with the License.
 # You may obtain a copy of the License at
 #
-#    http://www.apache.org/licenses/LICENSE-2.0
+# http://www.apache.org/licenses/LICENSE-2.0
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Nit, but can you revert this to the way it was?

File: bsuite/utils/gym_wrapper.py [hunk 3]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
   def __getattr__(self, attr):
     """Delegate attribute access to underlying environment."""
     return getattr(self._env, attr)
+
+
+class ReverseGymWrapper(dm_env.Environment):
+    """A wrapper that converts an OpenAI gym environment to a dm_env.Environment"""
+
+    def __init__(self, gym_env:gym.Env):
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Type annotations need a space after the ':' (ditto elsewhere)

File: bsuite/utils/gym_wrapper.py [hunk 4]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
...
+        self._observation_spec = self.space2spec(self.gym_env.observation_space, "observations")
+        self._action_spec = self.space2spec(self.gym_env.action_space, "actions")
+
+    def reset(self):
+        self.gym_env.reset()
+
+    def step(self, action):
+        """convert gym step result (observations, reward, done, info) to dm_env TimeStep"""
+        gym_step_res = self.gym_env.step(action)
+        _obs = gym_step_res[0]
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Why do these have _ prefixes?

File: bsuite/utils/gym_wrapper.py [hunk 5]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
...
+        self.gym_env = gym_env
+        # convert gym action and observation spaces to dm_env specs
+        self._observation_spec = self.space2spec(self.gym_env.observation_space, "observations")
+        self._action_spec = self.space2spec(self.gym_env.action_space, "actions")
+
+    def reset(self):
+        self.gym_env.reset()
+
+    def step(self, action):
+        """convert gym step result (observations, reward, done, info) to dm_env TimeStep"""
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Can you make your docstrings proper sentences, i.e. 'Convert .... TimeStep.' (ditto elsewhere)

File: bsuite/utils/gym_wrapper.py [hunk 6]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
...
+        gym_step_res = self.gym_env.step(action)
+        _obs = gym_step_res[0]
+        _reward = gym_step_res[1]
+        _done = gym_step_res[2]
+
+        if _done:
+            return dm_env.TimeStep(dm_env.StepType.LAST, _reward, None, _obs)
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Prefer to use dm_env.termination and dm_env.transition

File: bsuite/utils/gym_wrapper.py [hunk 7]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
...
+    def close(self):
+        self.gym_env.close()
+
+    def observation_spec(self):
+        return self._observation_spec
+
+    def action_spec(self):
+        return self._action_spec
+
+    def space2spec(self, space:gym.Space, name:str=None):
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Nit, this needs to be Text from the typing module for Python 2/3 compatibility.

File: bsuite/utils/gym_wrapper.py [hunk 8]
```diff
@@ -93,3 +94,77 @@ def reward_range(self) -> Tuple[float, float]:
...
+        elif isinstance(space, spaces.Dict):
+
+            spec_dict = {}
+            for k in space.spaces:
+                spec_dict[k] = self.space2spec(space.spaces[k], name)
+
+            return spec_dict
+
+        else:
+            raise ValueError("Unexpected gym space type")
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Needs a new line at end of file.

File: bsuite/utils/wrappers_test.py [hunk 9]
```diff
@@ -22,13 +22,14 @@
...
 from bsuite.utils import wrappers
+from bsuite.utils import gym_wrapper
 
 import dm_env
 from dm_env import specs
 from dm_env import test_utils
 import mock
 import numpy as np
-
+import gym
```
Comment (aslanides) [inline bsuite/utils/wrappers_test.py]: Imports should be alphabetical, so this should be after dm_env and before mock.

File: bsuite/utils/wrappers_test.py [hunk 10]
```diff
@@ -22,13 +22,14 @@
 
 from bsuite.experiments.catch import catch
 from bsuite.utils import wrappers
+from bsuite.utils import gym_wrapper
```
Comment (aslanides) [inline bsuite/utils/wrappers_test.py]: Imports should be alphabetical, so this should be after catch and before wrappers.

File: bsuite/utils/wrappers_test.py [hunk 11]
```diff
@@ -137,5 +138,33 @@ def make_action_sequence(self):
...
 
 
+class ReverseGymWrapperTest(absltest.TestCase):
+
+  def test_gym_cartpole(self):
+    self.gym_test_env = gym_wrapper.ReverseGymWrapper(gym.make('CartPole-v0'))
+
+    # test converted observation spec
+    self.assertEqual(type(self.gym_test_env.observation_spec()), specs.BoundedArray)
+    self.assertEqual(self.gym_test_env.observation_spec().shape,(4,))
```
Comment (aslanides) [inline bsuite/utils/wrappers_test.py]: Can you add a single space after commas?

File: bsuite/utils/gym_wrapper.py [hunk 12]
```diff
@@ -59,6 +59,7 @@ def render(self, mode: Text = 'rgb_array') -> Union[np.ndarray, bool]:
     if mode == 'human':
       if self.viewer is None:
         from gym.envs.classic_control import rendering  # pylint: disable=g-import-not-at-top
+
```
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Nit, revert unrelated change.

File: bsuite/utils/gym_wrapper.py (hunk 7 above)
Comment (aslanides) [inline bsuite/utils/gym_wrapper.py]: Does this need to be a public method?

Comment (yotam) [reply bsuite/utils/gym_wrapper.py]: It should also be a function, rather than a method.

File: bsuite/utils/wrappers_test.py [hunk 13]
```diff
@@ -137,5 +138,33 @@ def make_action_sequence(self):
       yield rng.choice(actions)
 
 
+class ReverseGymWrapperTest(absltest.TestCase):
+
+  def test_gym_cartpole(self):
+    self.gym_test_env = gym_wrapper.ReverseGymWrapper(gym.make('CartPole-v0'))
```
Comment (yotam) [inline bsuite/utils/wrappers_test.py]: There's no need to assign anything to `self` here. Can we just have a local `gym_env` variable?

File: bsuite/utils/gym_wrapper.py (hunk 4 above)
Comment (aslanides) [reply bsuite/utils/gym_wrapper.py]: Also it's more idiomatic to do something like

obs, reward, done, _ = self.gym_env.step(action)

File: bsuite/utils/gym_wrapper.py (hunk 8 above)
Comment (aslanides) [reply bsuite/utils/gym_wrapper.py]: Also we could make this error message better, e.g.

ValueError('Unexpected gym space: {}'.format(space))

Note also the single quotes ' ' for consistency with the rest of the file.

**Comment by alexminnaar** [comment]
Thanks for the comments @aslanides and @yotam , I should be able to make the corrections in the next day or so.

**Comment by alexminnaar** [comment]
@yotam @aslanides I made some changes based on your suggestions

**Comment by yotam** [comment]
Thanks for the changes Alex. I'm going to merge the change internally then push it and close here. Will make a few further edits and formatting changes on the way.

**Comment by yotam** [comment]
Merged in af5b3fe

**Comment by alexminnaar** [comment]
Thanks @yotam
